### PR TITLE
Lambda-promtail: Add support for processing SQS messages, add promtailClient Type, add logger, upgrade dependencies and fix unexpected flushing behaviors

### PR DESCRIPTION
**Pull request by CCOLLOT**
**What this PR does / why we need it**:

## Context
I want to :
+ Use lambda-promtail to ingest AWS VPC flow logs from S3 files, deploying it as an AWS lambda function.
+ Configure the s3 Notification to send messages to SQS instead of directly to lambda in order to be able to spin up a secondary SQS queue to be used as a dead-letter queue lambda can send unsuccessfully processed events to.
    + This way I am able to use the `AWS SQS redrive` feature to process logs after fixing an outage in the ingestion flow and I am not limited by lambda giving up after 2 (default) retries.

I faced several issues:
+ lambda-promtail tries to flush a batch when it gets bigger than `BATCH_SIZE`. The current implementation does not reset the size of the batch, leading to a flush to loki/promtail for every single added line after reaching `BATCH_SIZE` (and making the below file descriptor issue much worse because the `http.Client` is currently created in the `send` function.
+ lambda-promtail reached the quota of file_descriptors of AWS lambda (1024) very quickly due to http.Client being created thousands of times when processing large files (a VPC flow log file can reach dozens of MB very quick, meaning hundreds of thousands of lines.

## Added Feature
+ Add support for processing events from an SQS queue
     + Side effect: add support for on-failure recovery using and SQS queue.
     + If merged, I would happily upload my terraform sample code to setup the S3 --> SQS main--> lambda --> SQS DLQ -->redrive to SQS main

## Fixed bugs

+ The `flushBatch` function would reset the stream after flushing but it wouldn't reset the size of the batch, leading to lambda-promtail flushing every single following line to `loki/promtail`. In my case it led to logs being processed very slowly by lambda when the total size is way above the default `batchSize`. I would reach the 15 min timeout on AWS lambda for almost every single file.
+ The `http.Client`used to flush logs to `loki/promtail` was being recreated thousands of times, which made the process quickly run out of file descriptors (limited to 1024 in AWS lambda). I set it as a global variable to avoid impacting other parts of the code for now but if required we could do it a little cleaner via dependency injection.

After fixing those bugs my average lambda execution time for vpc flow log files (size between 10mb and 100mb) went from timing out at 15 min to ~3 seconds.

## Upgrades:
+ Go version to `1.19`
+ Dependencies using `go get -u ./...`

## Other
+ Reorganized imports 
+ Fixed some typos

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md`
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: tools/lambda-promtail/lambda-promtail/main.go
Comment (cstyan) [inline tools/lambda-promtail/lambda-promtail/main.go:179]: is this behaviour documented somewhere?

File: tools/lambda-promtail/lambda-promtail/main_test.go
Comment (cstyan) [inline tools/lambda-promtail/lambda-promtail/main_test.go:5]: can we avoid whitespace/import reordering changes, just makes reviews for new features easier :) thank you

File: tools/lambda-promtail/lambda-promtail/promtail.go
Comment (cstyan) [inline tools/lambda-promtail/lambda-promtail/promtail.go:132]: what is this for? fixing one of the bugs you're referring to I assume?

File: tools/lambda-promtail/lambda-promtail/s3.go
Comment (cstyan) [inline tools/lambda-promtail/lambda-promtail/s3.go:55]: did you mean to include this println?

**Review by cstyan — COMMENTED** [review]
few minor comments

reviewing this I also started thinking  that it might make sense to implement a `client` type within lambda promtail that we could attach functions to, and it could store the `promtailClient` reference rather than creating more package level globals

File: tools/lambda-promtail/lambda-promtail/s3.go
Comment (CCOLLOT) [reply tools/lambda-promtail/lambda-promtail/s3.go:55]: I used it for debugging purposes and ended up leaving it there since it's a useful piece of information (it helps immediately know what event/file is being processed) and because it prints only 1 line per lambda invocation (so it's not like it's going to cost a lot in Cloudwatch logs).

As the code is growing maybe we should have a more structured approach.
Maybe we could remove this line and open up another MR to use the standard logger (and set the LOG_LEVEL or DEBUG using an environment variable)? WDYT?

File: tools/lambda-promtail/lambda-promtail/promtail.go
Comment (CCOLLOT) [reply tools/lambda-promtail/lambda-promtail/promtail.go:132]: > what is this for? fixing one of the bugs you're referring to I assume?

Yes because batch.size is incremented as we scan the file and after each added line we compare batch.size and (max)batchSize to know if we need to flush the batch to the target (loki/promtail). 

It needs to be reset (to 0) after batch.size gets bigger than batchSize otherwise we end up sending a request to the target to flush a single line for each new line (so it becomes super slow because of the thousands of round-trip calls)

I think I will update this PR to provide a proper reset() method to make it clearer

File: tools/lambda-promtail/lambda-promtail/main.go
Comment (CCOLLOT) [reply tools/lambda-promtail/lambda-promtail/main.go:179]: Yes [here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-content-structure.html#:~:text=Example%20messages,following%20test%20message.)

**Comment by CCOLLOT** [comment]
Thanks! I will update this PR with your comment. 
Would you like me to implement the client type in this PR ?

File: tools/lambda-promtail/lambda-promtail/main_test.go
Comment (CCOLLOT) [reply tools/lambda-promtail/lambda-promtail/main_test.go:5]: Sure I will remove them 😅

**Comment by CCOLLOT** [comment]
Thanks! I will update this PR with your comments (and the new client type)

File: tools/lambda-promtail/lambda-promtail/main_test.go
Comment (cstyan) [reply tools/lambda-promtail/lambda-promtail/main_test.go:5]: thanks!

whitespace/import cleanup PRs are welcome! I just want to be sure I can easily see everything relevant to the things you're trying to fix here with this PR :)

File: tools/lambda-promtail/lambda-promtail/main.go
Comment (cstyan) [reply tools/lambda-promtail/lambda-promtail/main.go:179]: can we include that link in this comment please

File: tools/lambda-promtail/lambda-promtail/promtail.go
Comment (cstyan) [reply tools/lambda-promtail/lambda-promtail/promtail.go:132]: a `resetBatch` function sounds like a good idea :+1:

File: tools/lambda-promtail/lambda-promtail/s3.go
Comment (cstyan) [reply tools/lambda-promtail/lambda-promtail/s3.go:55]: I'm fine with keeping as is or implementing something more structured.

Comment (CCOLLOT) [reply tools/lambda-promtail/lambda-promtail/s3.go]: Okay I implemented the logger in my latest commit, I went with go-kit since it's the one used in the main project

File: tools/lambda-promtail/lambda-promtail/promtail.go
Comment (CCOLLOT) [reply tools/lambda-promtail/lambda-promtail/promtail.go:139]: Done in my latest commit 👍

**Comment by cstyan** [comment]
Looks like there's a conflict with your CHANGELOD.md changes. Might be easiest to remove that commit, merge upstream main into your branch, and then make your changelog entry again.  I'll have a look at the client and logger changes in the morning.

File: tools/lambda-promtail/lambda-promtail/promtail_client.go
Comment (cstyan) [inline tools/lambda-promtail/lambda-promtail/promtail_client.go:14]: we can just name the interface `client`

File: tools/lambda-promtail/lambda-promtail/s3.go
Comment (cstyan) [inline tools/lambda-promtail/lambda-promtail/s3.go:64]: why are we removing the object fetching code here and moving it?

File: tools/lambda-promtail/lambda-promtail/promtail_client.go
Comment (CCOLLOT) [reply tools/lambda-promtail/lambda-promtail/promtail_client.go:14]: Sure, will do! 👍  Actually I thought that it might get confusing later because there are other clients (s3clients etc...) that could be mocked and passed around using interfaces.  I guess it's not an issue to call this interface `client` for now and we still can move the clients to dedicated packages later on.

File: docs/sources/clients/lambda-promtail/_index.md
Comment (JStickler) [inline docs/sources/clients/lambda-promtail/_index.md:113]:
```suggestion
For AWS services supporting sending messages to SQS (for example, S3 with an S3 Notification to SQS), events can be processed through an `SQS queue with lambda trigger` instead of directly configuring the source service to trigger lambda. Lambda-promtail will retrieve the nested events from the SQS messages' body and process them as if them came directly from the source service.
```

File: docs/sources/clients/lambda-promtail/_index.md
Comment (JStickler) [inline docs/sources/clients/lambda-promtail/_index.md:116]:
```suggestion
Triggering lambda-promtail through SQS allows handling on-failure recovery of the logs using a secondary SQS queue as a dead-letter-queue (DLQ). You can configure lambda so that unsuccessfully processed messages will be sent to the DLQ. After fixing the issue, operators will be able to reprocess the messages by sending back messages from the DLQ to the source queue using the SQS `DLQ redrive` feature.
```

**Review by JStickler — COMMENTED** [review]
[Docs squad]  Couple of small suggestions.

File: docs/sources/clients/lambda-promtail/_index.md
Comment (CCOLLOT) [reply docs/sources/clients/lambda-promtail/_index.md:113]: Thanks for your suggestion! I will include it in the PR

File: docs/sources/clients/lambda-promtail/_index.md
Comment (CCOLLOT) [reply docs/sources/clients/lambda-promtail/_index.md:116]: Thanks for your suggestion! I will include it in the PR

Comment (CCOLLOT) [reply docs/sources/clients/lambda-promtail/_index.md]: Done, also added some links to the AWS documentation

**Review by JStickler — APPROVED** [review]
[docs sqad]  docs look good, I'll leave it to engineering to approve the code.

File: tools/lambda-promtail/lambda-promtail/s3.go
Comment (cstyan) [reply tools/lambda-promtail/lambda-promtail/s3.go]: :+1: got it

**Comment by cstyan** [comment]
I appreciate the bug fixes :+1: 

Having never used SQS myself I can't speak to the validity of the SQS specific code or how you're talking about using it, but I trust that you or other users will be using it enough to point out any issues that need fixing going forward.

I'll leave this for a day just to have one last read through and see if anyone else has comments, and then merge.

**Comment by CCOLLOT** [comment]
> I appreciate the bug fixes :+1: 
> 
> Having never used SQS myself I can't speak to the validity of the SQS specific code or how you're talking about using it, but I trust that you or other users will be using it enough to point out any issues that need fixing going forward.
> 
> I'll leave this for a day just to have one last read through and see if anyone else has comments, and then merge.

Glad I could help. I've been running this code for days/weeks now and it runs smoothly. 

Thanks for the review. ✌️

**Comment by cstyan** [comment]
hey @CCOLLOT my bad for leaving this unmerged, do you mind just resolving the conflicts? I will merge after that

**Review by cstyan — APPROVED** [review]

**Pull request merged**
